Support RPC 0.8.0

franciszekjob opened the pull request:
<!-- Reference any GitHub issues resolved by this PR -->

Towards #2490 

## Introduced changes

<!-- A brief description of the changes -->

Bump RPC to 0.8.0:
- Introduce `--l1-gas`, `--l1-gas-price`, `--l2-gas`, `--l2-gas-price`, `--l1-data`, `--l1-data-gas` flags
- Remove `--max-gas`, `--max-gas-unit-price` flags
- Update RPC errors

## Checklist

<!-- Make sure all of these are complete -->

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

franciszekjob committed:
Remove `sign_deployment_v1` for braavos

franciszekjob committed:
Use deploy v3 in account generation

franciszekjob committed:
Adjust `wait_for_tx`

franciszekjob committed:
Add todos

franciszekjob committed:
Adjust `SNCastStarknetError`

franciszekjob committed:
Adjust `build_transaction_status_response`

franciszekjob committed:
Update docs for fee-related flags

franciszekjob committed:
Add l1, l2, l1 data gas flags and fields

franciszekjob committed:
Update tests

franciszekjob committed:
Update docs

franciszekjob committed:
Update fee-relaed logic

franciszekjob committed:
Update rpc url in CI

franciszekjob committed:
Update starknet-rs dependency

franciszekjob committed:
Merge branch 'master' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-rpc-0.8.0

franciszekjob committed:
Partially fix linting

franciszekjob committed:
Add temporary devnet action

franciszekjob committed:
Fix `DEVNET_INSTALL_DIR`

franciszekjob committed:
Update devnet action

franciszekjob committed:
Add step with cloning devnet

franciszekjob committed:
Set `CACHE_ON_FAILURE` to true

franciszekjob committed:
Update devnet installation script; Remove composite action

franciszekjob committed:
Update devnet download url in installation script

franciszekjob committed:
Update devnet installation script

franciszekjob committed:
Disable Windows tests on CI

franciszekjob committed:
Fix invoke tests

franciszekjob committed:
Fix declare tests

franciszekjob committed:
Fix deploy tests

franciszekjob committed:
Add `EntryPointNotFound`

franciszekjob committed:
Ignore `should_dedup_urls` test

franciszekjob committed:
Fix call tests

franciszekjob committed:
Fix multicall tests

franciszekjob committed:
Fix and adjust account deploy tests

franciszekjob committed:
Ignore argent case in `test_happy_case_keystore` test for account create

franciszekjob committed:
Fix `test_wait_for_reverted_transaction`

franciszekjob committed:
Fix fixtures

franciszekjob committed:
Update `Cargo.lock`

franciszekjob committed:
Merge branch 'master' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-rpc-0.8.0

franciszekjob committed:
Refactor `try_into_fee_settings`

franciszekjob committed:
Fix linting

franciszekjob committed:
Fix linting

franciszekjob committed:
Refactor `invoke_contract`

franciszekjob committed:
Fix tests

franciszekjob committed:
Add caches for used rpc url

franciszekjob committed:
Temporarily ignore fork warning tests

franciszekjob committed:
Add `ensure_max_fee_greater_than_gas_values`

franciszekjob committed:
Fix linting

franciszekjob committed:
Fix `test_max_fee_less_than_resource_bounds_value`

franciszekjob committed:
Fix `test_all_args`

franciszekjob committed:
CI cleanup

franciszekjob committed:
Add todo

franciszekjob committed:
Code cleanup

franciszekjob committed:
Change `InsufficientMaxFee` to `InsufficientResourcesForValidate` in docs

franciszekjob committed:
Fix `test_happy_case_shell` deploy

franciszekjob committed:
Merge branch 'master' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-rpc-0.8.0

franciszekjob committed:
Update docs

franciszekjob committed:
Update changelog

franciszekjob committed:
Make self review

franciszekjob committed:
Add todos

franciszekjob committed:
Add todos

franciszekjob committed:
Add todos

franciszekjob committed:
Add todo

franciszekjob committed:
Update todo

franciszekjob committed:
Use `u128` in `mint_token`

franciszekjob committed:
Use starknet-rs crate

franciszekjob committed:
Restore rpc version in fork tests

franciszekjob commented on crates/sncast/tests/e2e/declare.rs:
```diff
@@ -82,8 +85,18 @@ async fn test_happy_case(class_hash: Felt, account_type: AccountType) {
         "Map",
-        "--max-fee",
-        "99999999999999999",
+        "--l1-gas",
+        "100000",
+        "--l1-gas-price",
+        "10000000000000",
+        "--l2-gas",
+        "1000000000",
+        "--l2-gas-price",
+        "100000000000000000000",
+        "--l1-data-gas",
+        "100000",
+        "--l1-data-gas-price",
+        "10000000000000",
```
note: Using auto estimation fails because fee estimate returns insufficient resources (it's on devnet side). Due to it, we use fixed resource bounds in our tests (applies to all other places).

franciszekjob commented:
note: Docs and scripts tests are failing, they will be fixed in subsequent PRs

ddoktorski commented on scripts/install_devnet.sh:
```diff
@@ -2,8 +2,9 @@
 DEVNET_INSTALL_DIR="$(git rev-parse --show-toplevel)/crates/sncast/tests/utils/devnet"
-DEVNET_REPO="https://github.com/0xSpaceShard/starknet-devnet-rs"
-DEVNET_VER="v0.2.2"
+# TODO(#3087): Use official devnet release
```
Why not [v0.3.0-rc.0 release](https://github.com/0xSpaceShard/starknet-devnet/releases/tag/v0.3.0-rc.0)?

franciszekjob replied:
It says:
> This is a pre-release serving JSON-RPC API v0.7.1 and internally executing transactions using blockifier v0.14.0-rc.0. This means it's mostly adapted to Starknet 0.13.4, with the exception of L1-data + L2 gas semantics, which are tied to JSON-RPC v0.8.0.

ddoktorski commented on crates/sncast/src/helpers/constants.rs:
```diff
@@ -20,7 +20,8 @@ pub const UDC_ADDRESS: Felt =
 pub const ARGENT_CLASS_HASH: Felt =
-    felt!("0x029927c8af6bccf3f6fda035981e765a7bdbf18a2dc0d630494f8758aa908e2b");
+    // felt!("0x029927c8af6bccf3f6fda035981e765a7bdbf18a2dc0d630494f8758aa908e2b");
+    felt!("0x036078334509b514626504edc9fb252328d1a240e4e948bef8d0c08dff45927f");
```
Why this class hash?

franciszekjob replied:
I've restored it to previous version 0(.3.1). However some tests fail with it (there's no such problem with 0.4.0 which was present the change above). Will be addressed in #3089

franciszekjob replied:
Yup.

ddoktorski commented on crates/sncast/src/helpers/fee.rs:
```diff
@@ -1,117 +1,174 @@
+        // but in case someone passes --max-fee flag, we need to make estimation on our own
+        // to check if the fee estimate isn't higher than provided max fee
+        if let Some(max_fee) = self.max_fee {
+            self.check_conflicting_flags()?;
```
It should be validated by clap

franciszekjob replied:
Not sure if that's possible  on clap level- we want to disallow passing all together, but still allow to pass any other subset. `conflicts_with_all` seemd helpful, but after all it doesn't allow for that.

cptartur replied:
I don't think we need this kind of level of granularity of settings fees: For me user should be either allowed to pass `--max-fee` or specific args for gas. I don't see use cases for setting max fee along specific gas values in a command line too.

This would make it far less complex.

cptartur replied:
And we could validate everything at clap level.

cptartur replied:
Something along the lines of this

```
// Arg group "resources"
    pub l1_gas: Felt,
    pub l1_gas_price: Felt,
    pub l2_gas: Felt,
    pub l2_gas_price: Felt,
    pub l1_data_gas: Felt
    pub l1_data_gas_price: Felt,


// FeeArgs
max_fee: Felt // conflicts with arg group resources
resources: Resources // conflicts with max fee
```

franciszekjob replied:
> I don't think we need this kind of level of granularity of settings fees: For me user should be either allowed to pass > --max-fee or specific args for gas. I don't see use cases for setting max fee along specific gas values in a > command line too.

> This would make it far less complex.

My initial idea was to allow using max fee with other args, but if we want to simplify it then that's fine.

ddoktorski commented on crates/sncast/src/helpers/fee.rs:
```diff
@@ -1,117 +1,174 @@
+            self.check_conflicting_flags()?;
+            self.validate_max_fee_meets_resource_bounds()?;
+
+            ensure!(
+                fee_estimate.is_some(),
+                "Fee estimate is required when passing max fee is provided"
+            );
```
```suggestion
            let fee_estimate = fee_estimate.clone().ok_or_else(|| {
                anyhow!("Fee estimate must be calculated when max_fee is provided")
            })?;
```

franciszekjob committed:
Use `TEST_RESOURCE_BOUNDS_FLAGS`

ddoktorski committed:
Replace equilibrium rpc url with our internal node (#3094)

<!-- Reference any GitHub issues resolved by this PR -->

Closes #3086

---------

Co-authored-by: Fiiranek <jellybean@onet.eu>

kkawula committed:
Fix tests

franciszekjob commented on crates/sncast/src/response/errors.rs:
```diff
@@ -73,14 +75,14 @@ pub enum SNCastStarknetError {
     InvalidTransactionNonce,
-    #[error("Max fee is smaller than the minimal transaction cost")]
-    InsufficientMaxFee,
+    #[error("The transaction's resources don't cover validation or the minimal transaction fee")]
```
This is taken from [spec](https://github.com/starkware-libs/starknet-specs/blob/4b351773f92feb7a2b1b9a754ff525ee0ba194bb/api/starknet_write_api.json#L239) 😅 .

ddoktorski replied:
Up to you

franciszekjob commented on crates/sncast/src/starknet_commands/multicall/mod.rs:
```diff
@@ -15,6 +15,6 @@ pub struct Multicall {
 #[derive(Debug, Subcommand)]
 pub enum Commands {
-    Run(Run),
+    Run(Box<Run>),
```
It was suggested by clippy, but it's not needed anymore. Removed.

cptartur replied:
What's the lint name? Seems unusual.

franciszekjob replied:
https://rust-lang.github.io/rust-clippy/master/index.html#/large_enum_variant

Seems it's not needed now, as the size is ok.

cptartur replied:
TIL, okay, let's keep that

franciszekjob committed:
Apply code review suggestions

franciszekjob commented on crates/sncast/src/helpers/fee.rs:
```diff
@@ -1,117 +1,174 @@
-                        NonZeroU128::try_from_(max_gas_unit_price).map_err(anyhow::Error::msg)?,
-                    ),
-                }
+    pub fn try_into_fee_settings(&self, fee_estimate: &Option<FeeEstimate>) -> Result<FeeSettings> {
```
In fact, clippy enforces to use reference.

cptartur replied:
Let's use `Option<&FeeEstimate>` it's more idiomatic.

franciszekjob committed:
Use previous argent class hash

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-rpc-0.8.0

franciszekjob committed:
Ignore block explorer tests for nftscan

franciszekjob committed:
Merge branch 'ignore-block-explorer-nftscan-tests' into kkawula/fork-warning-tests

franciszekjob committed:
Merge branch 'ignore-block-explorer-nftscan-tests' into franciszekjob/2490-rpc-0.8.0

franciszekjob committed:
Merge branch 'kkawula/fork-warning-tests' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-rpc-0.8.0

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-rpc-0.8.0

franciszekjob committed:
Fix fee args validation

franciszekjob committed:
Add todos

franciszekjob committed:
Use zero as guardian key

franciszekjob committed:
Merge branch 'master' into franciszekjob/2490-rpc-0.8.0

cptartur commented on crates/conversions/src/serde/serialize/serialize_impl.rs:
```diff
@@ -27,14 +29,28 @@ impl CairoSerialize for CallEntryPoint {
+    }
+}
+
+impl CairoSerialize for ContractExecutionError {
+    fn serialize(&self, output: &mut BufferWriter) {
+        match &self {
+            ContractExecutionError::Nested(inner) => {
+                inner.class_hash.serialize(output);
+                inner.contract_address.serialize(output);
+                inner.selector.serialize(output);
+                inner.error.serialize(output);
+            }
+            ContractExecutionError::Message(msg) => ByteArray::from(msg.as_str()).serialize(output),
+        }
```
Do we actually test these anywhere? I think it would be worth doing so.

franciszekjob replied:
I made output assertion in `test_wrong_function_name` and `test_wrong_calldata` more detailed, so now it checks the serialization, take a look.

cptartur replied:
Okay, ideally we'd have some tests here as well but okay.

cptartur commented on crates/forge/tests/e2e/fork_warning.rs:
```diff
@@ -39,6 +39,7 @@ fn setup_redirect_server() {
 }
 
 #[test]
+#[ignore = "TODO(#3088)"]
```
The issue for this is closed, shouldn't these be restored?

cptartur commented on crates/sncast/src/helpers/fee.rs:
```diff
@@ -120,27 +188,22 @@ impl FeeArgs {
 #[derive(Debug, PartialEq, CairoDeserialize)]
 pub struct ScriptFeeSettings {
```
At cast script level this should take an enum
```
pub enum ScriptFeeSettings {
  MaxFee(value),
  GasTriplet(...)
}
```

That's a feature for future versions but please create an issue for that.

cptartur commented on crates/sncast/src/lib.rs:
```diff
@@ -591,11 +596,11 @@ pub async fn wait_for_tx(
-                starknet::core::types::TransactionExecutionStatus::Reverted => {
-                    return get_revert_reason(provider, tx_hash).await;
+                starknet::core::types::ExecutionResult::Reverted { reason } => {
+                    return Ok(reason);
```
This isn't correct, but is being resolved by cast scripts PR.

cptartur commented on crates/sncast/src/starknet_commands/declare.rs:
```diff
@@ -78,24 +73,39 @@ pub async fn declare(
+        l1_data_gas_price,
+    } = fee_settings.expect("Failed to convert to fee settings");
+
+    let declaration = apply_optional(declaration, l1_gas, DeclarationV3::l1_gas);
+    let declaration = apply_optional(declaration, l1_gas_price, DeclarationV3::l1_gas_price);
+    let declaration = apply_optional(declaration, l2_gas, DeclarationV3::l2_gas);
+    let declaration = apply_optional(declaration, l2_gas_price, DeclarationV3::l2_gas_price);
+    let declaration = apply_optional(declaration, l1_data_gas, DeclarationV3::l1_data_gas);
```
This logic is repeated in multiple places, can't this be extracted to some generic helper function?

cptartur replied:
Or a macro if that's needed

franciszekjob replied:
It can, sure - added a macro for that.

cptartur commented on crates/sncast/src/starknet_commands/multicall/run.rs:
```diff
@@ -56,7 +56,7 @@ struct InvokeCall {
 
 pub async fn run(
-    run: Run,
+    run: Box<Run>,
```
Same here

cptartur commented on crates/sncast/tests/e2e/account/deploy.rs:
```diff
@@ -38,11 +40,10 @@ pub async fn test_happy_case(class_hash: &str, account_type: &str) {
-        "100000000000",
-    ];
+    ]
+    .into_iter()
+    .chain(TEST_RESOURCE_BOUNDS_FLAGS.into_iter())
+    .collect::<Vec<&str>>();
```
Nit, but maybe this could be hidden behind some helper function?

franciszekjob replied:
Sure, this should be better:

```rust
let args = vec![
        "--accounts-file",
        "accounts.json",
        "--account",
        "my_account",
        "deploy",
        "--url",
        URL,
        "--class-hash",
        MAP_CONTRACT_CLASS_HASH_SEPOLIA,
        "--salt",
        "0x2",
        "--unique",
    ];
    let args = apply_test_resource_bounds_flags(args);
```

cptartur commented on crates/sncast/tests/e2e/account/deploy.rs:
```diff
@@ -217,34 +190,34 @@ pub async fn test_valid_class_hash() {
-        transaction: [..]
-    "});
-}
+// #[tokio::test]
+// pub async fn test_valid_no_max_fee() {
+//     let tempdir = create_account(true, &OZ_CLASS_HASH.into_hex_string(), "oz").await;
+//     let accounts_file = "accounts.json";
+
+//     let args = vec![
+//         "--profile",
+//         "deploy_profile",
+//         "--accounts-file",
+//         accounts_file,
+//         "account",
+//         "deploy",
+//         "--url",
+//         URL,
+//         "--name",
+//         "my_account",
+//     ];
+
+//     let snapbox = runner(&args).current_dir(tempdir.path());
+
+//     snapbox.assert().success().stdout_matches(indoc! {r"
+//         command: account deploy
+//         transaction_hash: [..]
+
+//         To see invocation details, visit:
+//         transaction: [..]
+//     "});
+// }
```
Please restore it or ignore it with a linked issue

cptartur commented on crates/sncast/tests/e2e/call.rs:
```diff
@@ -160,11 +169,12 @@ fn test_wrong_calldata() {
         indoc! {r"
         command: call
-        error: An error occurred [..]Execution failed[..]Input too long for arguments[..]
+        response: [0x496e70757420746f6f206c6f6e6720666f7220617267756d656e7473]
```
Why isn't it decoded now?

franciszekjob replied:
It's coming from devnet (which gets that from blockifier). Previously it was returned as error, now it's a response (not sure why, it's not because of our codebase). Create #3107 for that.

cptartur commented on crates/sncast/tests/e2e/declare.rs:
```diff
@@ -95,68 +100,69 @@ async fn test_happy_case(class_hash: Felt, account_type: AccountType) {
-    assert!(matches!(receipt, Declare(_)));
-}
+// TODO(#3090)
+// #[test_case(Some("100000000000000000"), None, None; "max_fee")]
```
I think that's a big one to actually ignore, I'd prefer we restore it in some form that before we release.

cptartur replied:
Also let's uncomment this and ignore instead of making it commented out.

franciszekjob replied:
I restored it with resource bounds case, the one with max fee is ignored (linked to #3100).

cptartur commented on crates/sncast/tests/e2e/deploy.rs:
```diff
@@ -91,56 +94,57 @@ async fn test_happy_case(class_hash: Felt, account_type: AccountType) {
-
-    assert!(matches!(receipt, Deploy(_)));
-}
+// TODO(#3090)
```
Uncomment and ignore

franciszekjob replied:
Refactored it and linked #3100 . The test case with max fee has been commented out as it's not possible to ignore a specific one.

cptartur commented on crates/sncast/tests/e2e/invoke.rs:
```diff
@@ -91,67 +97,61 @@ async fn test_happy_case(class_hash: Felt, account_type: AccountType) {
-
-    assert!(matches!(receipt, Invoke(_)));
-}
+// TODO(#3090)
```
Uncomment and ignore

franciszekjob replied:
Refactored it and linked #3100 . The test case with max fee has been commented out as it's not possible to ignore a specific one.

cptartur commented on crates/sncast/tests/e2e/invoke.rs:
```diff
@@ -267,78 +275,45 @@ fn test_too_low_gas() {
-        "},
-    );
-}
+// #[test]
+// fn test_calculated_max_gas_equal_to_zero_when_max_fee_passed() {
+//     let args = vec![
+//         "--accounts-file",
+//         ACCOUNT_FILE_PATH,
+//         "--account",
+//         "user11",
+//         "--wait",
+//         "invoke",
+//         "--url",
+//         URL,
+//         "--contract-address",
+//         MAP_CONTRACT_ADDRESS_SEPOLIA,
+//         "--function",
+//         "put",
+//         "--calldata",
+//         "0x1",
+//         "0x2",
+//         "--max-fee",
+//         "999999",
+//     ];
+
+//     let snapbox = runner(&args);
+//     let output = snapbox.assert().success();
+
+//     // TODO(#2852)
+//     assert_stderr_contains(
+//         output,
+//         indoc! {r"
+//         command: invoke
+//         error: Calculated max-gas from provided --max-fee and the current network gas price is 0. Please increase --max-fee to obtain a positive gas amount: Tried to create NonZeroFelt from 0
+//         "},
+//     );
+// }
```
Uncomment (and ignore with a linked issue if needed). Or remove if it is no longer relevant (kind of seems like it).

cptartur commented on crates/sncast/tests/helpers/constants.rs:
```diff
@@ -23,7 +24,7 @@ pub const DEVNET_PREDEPLOYED_ACCOUNT_ADDRESS: &str =
 // OpenZeppelin account contract v0.8.1
 pub const DEVNET_OZ_CLASS_HASH_CAIRO_1: Felt =
-    felt!("0x061dac032f228abef9c6626f995015233097ae253a7f72d68552db02f2971b8f");
+    felt!("0x044cab2e6a3a7bc516425d06d76c6ffd56ae308864dbc66f8e75028e3784aa29");
```
Class hash changed but the comment above says the version is the same? Why is that?

franciszekjob replied:
The comment with version should be changed, but tbh I don't think it's worth keeping at all, we don't do that for Cairo 0, nor for argent or braavos. Wdyt?

cptartur replied:
Okay, we can remove it, better than provide incorrect informations.

ddoktorski replied:
* Are you sure this is the correct class hash? In the devnet repo it is named [sierra hash](https://github.com/0xSpaceShard/starknet-devnet/blob/cb779e696c07c89198a64c4d9c3ffde46bc7a66d/tests/integration/common/constants.rs#L26) not class hash
* Why is it `felt` instead of `&str`?
* You can find official calculated OZ class hashes in https://docs.openzeppelin.com/contracts-cairo/1.0.0/presets

franciszekjob replied:
> Are you sure this is the correct class hash?

Yes, it's the one used for oz accounts predeployement:
![image](https://github.com/user-attachments/assets/a34b74b0-6a19-4d6e-bb3d-13b0fc49e660)

> Are you sure this is the correct class hash?
It's just the name on their side.

> Why is it felt instead of &str?
It's not a change I included in my PR.

> You can find official calculated OZ class hashes in https://docs.openzeppelin.com/contracts-cairo/1.0.0/presets

You have a good eye 👍

cptartur commented on crates/sncast/tests/helpers/fixtures.rs:
```diff
@@ -215,15 +224,37 @@ pub async fn invoke_contract(
 
-    let execution = account.execute_v1(vec![call]);
-    let execution = apply_optional(execution, max_fee, ExecutionV1::max_fee);
+    let execution = account.execute_v3(vec![call]);
+    let execution = apply_optional(execution, fee_settings.l1_gas, ExecutionV3::l1_gas);
+    let execution = apply_optional(
+        execution,
+        fee_settings.l1_gas_price,
+        ExecutionV3::l1_gas_price,
+    );
+    let execution = apply_optional(execution, fee_settings.l2_gas, ExecutionV3::l2_gas);
+    let execution = apply_optional(
+        execution,
+        fee_settings.l2_gas_price,
+        ExecutionV3::l2_gas_price,
+    );
+    let execution = apply_optional(
+        execution,
+        fee_settings.l1_data_gas,
+        ExecutionV3::l1_data_gas,
+    );
+    let execution = apply_optional(
+        execution,
+        fee_settings.l1_data_gas_price,
+        ExecutionV3::l1_data_gas_price,
+    );
```
If we create a helper for these it can be used here as well

franciszekjob replied:
Same as above, macro `apply_optional_fields` handles that now.

franciszekjob committed:
Apply code review suggestions

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-rpc-0.8.0

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-rpc-0.8.0

franciszekjob committed:
Add `apply_optional_fields` macro

franciszekjob committed:
Update `test_wrong_function_name` and `test_wrong_calldata`

franciszekjob committed:
Apply code review suggestions

franciszekjob committed:
Minor code cleanup

ddoktorski commented on crates/sncast/src/helpers/fee.rs:
```diff
@@ -1,146 +1,179 @@
+            );
 
-        Ok(settings)
+            let fee_settings = FeeSettings {
```
Why not to implement `TryFrom<FeeEstimate>` and `TryFrom<FeeArgs>` for `FeeSettings`?

ddoktorski commented on crates/sncast/src/starknet_commands/account/create.rs:
```diff
@@ -181,8 +181,9 @@ async fn generate_account(
         AccountType::Argent => {
             let factory =
-                ArgentAccountFactory::new(class_hash, chain_id, Felt::ZERO, signer, provider)
+                ArgentAccountFactory::new(class_hash, chain_id, Some(Felt::ZERO), signer, provider)
```
This is for a guardian public key. I guess they changed the interface and now it should be `None`, but please check this 🙏 
```suggestion
                ArgentAccountFactory::new(class_hash, chain_id, None, signer, provider)
```

ddoktorski replied:
Perhaps this is related to issues with argent accounts that you mentioned

franciszekjob replied:
> This is for a guardian public key. I guess they changed the interface and now it should be None, but please check this 🙏

Yes, in `v0.4.0` guardian key is optional but none that I've used `Some(Felt::ZERO)` because we're still using class hash for v0.3.1. However, I think we're good to update to v0.4.1 (which is used while creating account) so the guardian key will be `None` then. Updated.

ddoktorski commented on crates/sncast/src/starknet_commands/account/deploy.rs:
```diff
@@ -194,7 +194,7 @@ async fn get_deployment_result(
                 class_hash,
                 chain_id,
-                Felt::ZERO,
+                Some(Felt::ZERO),
```
Same as above

cptartur commented on crates/sncast/src/lib.rs:
```diff
@@ -731,6 +718,19 @@ pub fn apply_optional<T, R, F: FnOnce(T, R) -> T>(initial: T, option: Option<R>,
 }
 
+#[macro_export]
+macro_rules! apply_optional_fields {
```
No way apply_optional got a sequel

franciszekjob replied:
😆

cptartur commented on crates/sncast/src/starknet_commands/deploy.rs:
```diff
@@ -92,35 +92,45 @@ pub async fn deploy(
+        l1_data_gas_price => DeploymentV3::l1_data_gas_price,
+        nonce => DeploymentV3::nonce
+    );
+    // let execution = match l1_gas {
+    //     None => execution,
+    //     Some(l1_gas) => execution.l1_gas(l1_gas),
+    // };
+    // let execution = match l1_gas_price {
+    //     None => execution,
+    //     Some(l1_gas_price) => execution.l1_gas_price(l1_gas_price),
+    // };
+    // let execution = match l2_gas {
+    //     None => execution,
+    //     Some(l2_gas) => execution.l2_gas(l2_gas),
+    // };
+    // let execution = match l2_gas_price {
+    //     None => execution,
+    //     Some(l2_gas_price) => execution.l2_gas_price(l2_gas_price),
```
Please remove the commented out code

cptartur commented on crates/sncast/src/starknet_commands/deploy.rs:
```diff
@@ -92,35 +92,45 @@ pub async fn deploy(
-        Some(nonce) => execution.nonce(nonce),
-    };
-    let result = execution.send().await;
+    let deployment = apply_optional_fields!(
+        deployment,
+        l1_gas => DeploymentV3::l1_gas,
+        l1_gas_price => DeploymentV3::l1_gas_price,
+        l2_gas => DeploymentV3::l2_gas,
+        l2_gas_price => DeploymentV3::l2_gas_price,
+        l1_data_gas => DeploymentV3::l1_data_gas,
+        l1_data_gas_price => DeploymentV3::l1_data_gas_price,
+        nonce => DeploymentV3::nonce
```
Assuming that the commented out code was just condensed to that?

franciszekjob replied:
Yup, left the commented out one by accident.

cptartur commented on crates/sncast/tests/docs_snippets/validation.rs:
```diff
@@ -65,14 +64,6 @@ fn test_docs_snippets() {
-        {
-            args.extend(TEST_RESOURCE_BOUNDS_FLAGS);
-        }
-
```
Why isn't it required anymore?

franciszekjob replied:
This change should was accidentally added to this PR, it should stay on https://github.com/foundry-rs/starknet-foundry/pull/3101

franciszekjob committed:
Remove oz version comment

franciszekjob committed:
Add impls for `FeeSettings`

franciszekjob committed:
Apply code review suggestions

franciszekjob committed:
Update argent and braavos class hashes; Enable some tests for argent

franciszekjob committed:
Add adjustment to `EXPECTED_RPC_VERSION`

franciszekjob committed:
Adjust setting EXPECTED_RPC_VERSION

franciszekjob committed:
Update adjustment for EXPECTED_RPC_VERSION

franciszekjob committed:
Restore old braavos class hash

franciszekjob committed:
Update changelog

franciszekjob committed:
Uncomment some tests

franciszekjob committed:
Add todos

franciszekjob committed:
Fix tests

ddoktorski commented on crates/sncast/Cargo.toml:
```diff
@@ -30,7 +30,7 @@ runtime = { path = "../runtime" }
 data-transformer = { path = "../data-transformer" }
 configuration = { path = "../configuration" }
-shared = { path = "../shared" }
+shared = { path = "../shared", features = ["testing"] }
```
I think this way `testing` feature is always enabled. Is it the intended behavior?

franciszekjob replied:
Absolutely, changed.

ddoktorski commented on crates/sncast/src/helpers/fee.rs:
```diff
@@ -1,146 +1,193 @@
+                .expect("Failed to convert FeeEstimate to FeeSettings");
+            Ok(fee_settings)
+        } else {
+            let fee_settings = FeeSettings::try_from_(self.clone())
+                .expect("Failed to convert FeeArgs to FeeSettings");
```
We should not allow a panic scenario depending on user input. Please improve the error handling for this.

franciszekjob replied:
> In what cases can it even fail? Ideally it'd be verified on clap level or somewhere else and you'd just use ::from_ here.

Change to `from`.

> Wouldn't it make sense to use `u128` and `u64` accordingly in `FeeArgs` instead of typing these as felts? If we are converting them anyway.

Yes, I assume that before we had these values as felt because we wanted to accept hex strings. However, in spec they are defined as u64/u128, so I think we can safely change them.

ddoktorski commented on crates/sncast/src/helpers/fee.rs:
```diff
@@ -1,146 +1,193 @@
-    println!(
-        "Specifying '--max-fee' flag results in conversion to '--max-gas' and '--max-gas-unit-price' flags\nConverted {max_fee} max fee to {max_gas} max gas and {max_gas_unit_price} max gas unit price\n",
-    );
+impl TryFromConv<FeeEstimate> for FeeSettings {
```
Why `TryFromConv` instead of Rust `TryFrom`?

ddoktorski commented on crates/sncast/src/helpers/fee.rs:
```diff
@@ -1,146 +1,193 @@
-        "Specifying '--max-fee' flag results in conversion to '--max-gas' and '--max-gas-unit-price' flags\nConverted {max_fee} max fee to {max_gas} max gas and {max_gas_unit_price} max gas unit price\n",
-    );
+impl TryFromConv<FeeEstimate> for FeeSettings {
+    type Error = String;
```
Why string?

franciszekjob replied:
Change to anyhow error.

franciszekjob committed:
Apply code review suggestions

franciszekjob commented on crates/sncast/tests/e2e/deploy.rs:
```diff
@@ -302,7 +306,7 @@ fn test_contract_already_deployed() {
         indoc! {r"
         command: deploy
-        error: [..]Deployment failed: contract already deployed at address [..]
+        error: Transaction execution error [..]
```
I'm not sure what's the reason for that, have a look - below I've printed stderr for `test_contract_already_deployed` and `test_same_salt_and_class_hash_deployed_twice` (from scripts). We can clearly see that in script test there's additional message test.

![image](https://github.com/user-attachments/assets/6643b52d-5202-4192-bc87-41adb51dd9e0)

![image](https://github.com/user-attachments/assets/9730cb4f-5ec3-4a9c-8cc2-3ccd5393fc6a)

cptartur replied:
In scripts we also do additional print of the result from within the script code `println!("{:?}", deploy_result);`. 

Maybe there's something with out error handling/serialization logic that this get's omitted when running from the CLI?

franciszekjob replied:
Discussed offline: will be addresses in fixes before stable.

franciszekjob committed:
Merge branch 'master' into franciszekjob/2490-rpc-0.8.0

franciszekjob committed:
Update account deployment example

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-rpc-0.8.0

cptartur commented on crates/sncast/src/response/errors.rs:
```diff
@@ -73,14 +75,14 @@ pub enum SNCastStarknetError {
     InvalidTransactionNonce,
-    #[error("Max fee is smaller than the minimal transaction cost")]
-    InsufficientMaxFee,
+    #[error("The transaction's resources don't cover validation or the minimal transaction fee")]
+    InsufficientResourcesForValidate,
```
So there is no error for insufficient max fee now? Is it some kind of resources bounds error now?

franciszekjob replied:
Yes, when resource bounds are too low to validate tx, see [spec](https://github.com/starkware-libs/starknet-specs/blob/d70a2e57c9a66db1bbf86c388f38d6295a6a2a75/api/starknet_write_api.json#L42).

cptartur commented on crates/sncast/tests/integration/fee.rs:
We should have a test where estimate fee isn't none as well.

1. for testing if it get's correctly used
2. for testing if `max_fee` not being none and `estimate` being provided
3. for testing if `max_fee` being provided and `estimate` being none

Maybe some other?

franciszekjob replied:
1. `test_from_fee_estimate`
2. `test_max_fee_set`
3. `test_max_fee_set_and_fee_estimate_none`

> Maybe some other?
`test_max_fee_set_and_fee_estimate_higher`

franciszekjob committed:
Merge branch 'master' into franciszekjob/2490-rpc-0.8.0

franciszekjob committed:
Add todo

franciszekjob committed:
Add fee tests

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-rpc-0.8.0

franciszekjob committed:
Fix fork tests

cptartur commented on crates/sncast/tests/integration/fee.rs:
```diff
@@ -1,247 +1,264 @@
+        l1_gas_price: Felt::from(2),
+        l2_gas_consumed: Felt::from(3),
+        l2_gas_price: Felt::from(4),
+        l1_data_gas_consumed: Felt::from(5),
+        l1_data_gas_price: Felt::from(6),
+        unit: PriceUnit::Fri,
+        overall_fee: Felt::from(44),
     };
-
-    let settings = args
-        .try_into_fee_settings(factory.provider(), factory.block_id())
-        .await
-        .unwrap();
+    let settings = FeeSettings::try_from(mock_fee_estimate).unwrap();
 
     assert_eq!(
         settings,
         FeeSettings {
-            max_gas: None,
-            max_gas_unit_price: None,
+            l1_gas: Some(1),
+            l1_gas_price: Some(2),
+            l2_gas: Some(3),
+            l2_gas_price: Some(4),
+            l1_data_gas: Some(5),
+            l1_data_gas_price: Some(6),
         }
     );
```
This is basically a unit test, please place it directly in fee.rs file.

Maybe some other from here can be moved as well?

franciszekjob replied:
Good catch, moved.

> Maybe some other from here can be moved as well?

No, nothing apart of that.

cptartur commented on crates/sncast/src/helpers/fee.rs:
```diff
@@ -1,146 +1,183 @@
+        if let Some(max_fee) = self.max_fee {
+            self.validate_max_fee_meets_resource_bounds()?;
+
+            let fee_estimate = fee_estimate
+                .ok_or_else(|| anyhow!("Fee estimate must be passed when max_fee is provided"))?;
```
Actually, now that I look at it, it doesn't make sense to bubble it as a handleable error. If this error handling code is reached it means we have a bug in our logic somewhere else and this should never be allowed. Let's replace it with `expect()`.

cptartur commented on crates/sncast/src/helpers/fee.rs:
```diff
@@ -1,146 +1,183 @@
+    pub fn try_into_fee_settings(&self, fee_estimate: Option<&FeeEstimate>) -> Result<FeeSettings> {
+        // If some resource bounds values are lacking, starknet-rs will estimate them automatically
+        // but in case someone passes --max-fee flag, we need to make estimation on our own
+        // to check if the fee estimate isn't higher than provided max fee
```
Let's add an assert for case when `max_fee` is none and `fee estimate` is some. This means a bug in our logic and we should panic.

franciszekjob replied:
[This line](https://github.com/foundry-rs/starknet-foundry/blob/franciszekjob/2490-rpc-0.8.0/crates/sncast/src/helpers/fee.rs#L69-L70) checks it.

franciszekjob replied:
I don't think such assert is needed, as we already check if fee estimate is not none, inside first if's block.

franciszekjob committed:
Apply code review suggestions

cptartur commented on crates/sncast/tests/integration/fee.rs:
```diff
@@ -1,247 +1,238 @@
 
     assert_eq!(
-        settings,
-        FeeSettings {
-            max_gas: Some(NonZeroU64::new(MAX_FEE / 1_000).unwrap()),
-            max_gas_unit_price: Some(NonZeroU128::new(1_000).unwrap()),
-        }
+        err.to_string(),
+        "Fee estimate must be passed when max_fee is provided"
```
I think it should just be 

```rust
args.try_into_fee_settings(None).unwrap();
```

Without any other lines? If the tests is `should_panic` anyway.

franciszekjob committed:
Merge branch 'master' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-rpc-0.8.0

franciszekjob committed:
Fix changelog

franciszekjob committed:
Fix `test_max_fee_set_and_fee_estimate_none`

ddoktorski commented on docs/src/appendix/sncast/account/deploy.md:
```diff
@@ -23,14 +23,34 @@ Possible values: `mainnet`, `sepolia`.
 Optional.
 
-Maximum fee for the `deploy_account` transaction in Fri or Wei depending on fee token or transaction version. When not used, defaults to auto-estimation. Must be greater than zero.
+Maximum fee for the `deploy_account` transaction depending on fee token or transaction version. If used, estimated overall fee can't be higher than this value. Must be greater than zero.
```
```suggestion
Maximum fee for the `deploy_account` denoted in FRI. Must be greater than zero. If provided, it is not possible to use any of the following fee related flags: `--l1-gas`, `--l1-data-price`, `--l2-gas`, `--l2-gas-price`, `--l1-data-gas`, `--l1-data-gas-price`.
```
Also for other commands

franciszekjob commented on crates/sncast/src/helpers/fee.rs:
```diff
@@ -1,150 +1,222 @@
+        }
+    }
+
+    fn validate_max_fee_meets_resource_bounds(&self) -> Result<()> {
```
No, let's remove it.

franciszekjob committed:
Appply code review suggestion

franciszekjob committed:
Update sncast scripts (#3076)

<!-- Reference any GitHub issues resolved by this PR -->

Closes #3102
Towards #2490

## Introduced changes

<!-- A brief description of the changes -->

- Update `FeeSettings` struct
- Update and fix scripts tests

## Checklist

<!-- Make sure all of these are complete -->

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

---------

Co-authored-by: Artur Michałek <52135326+cptartur@users.noreply.github.com>

franciszekjob committed:
Disable braavos integration (#3122)

<!-- Reference any GitHub issues resolved by this PR -->

Closes #

## Introduced changes

<!-- A brief description of the changes -->

Temporarily disable integration with braavos accounts. Reason for that
described in #3118

## Checklist

<!-- Make sure all of these are complete -->

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

---------

Co-authored-by: Artur Michałek <52135326+cptartur@users.noreply.github.com>

franciszekjob committed:
Update braavos message

franciszekjob committed:
Remove `test_max_fee_less_than_resource_bounds_value`

franciszekjob committed:
Add `ContractExecutionError` (#3138) [skip ci]

<!-- Reference any GitHub issues resolved by this PR -->

Closes #3120 

## Introduced changes

<!-- A brief description of the changes -->

Refactor errors in `sncast_std` according to spec

## Checklist

<!-- Make sure all of these are complete -->

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

---------

Co-authored-by: Artur Michałek <52135326+cptartur@users.noreply.github.com>

franciszekjob committed:
Merge branch 'master' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-rpc-0.8.0

franciszekjob committed:
Temporarily enable braavos integration

franciszekjob committed:
Revert "Temporarily enable braavos integration"

This reverts commit 0a198cef08c05d4a79cfba9bdad9063f2c15736a.

franciszekjob committed:
Bump starknet-rs
